public/js/components/util/ManagedTree.js

```diff
@@ -0,0 +1,58 @@
+"use strict";
+
+const React = require("react");
+const Tree = React.createFactory(require("../../lib/tree"));
+
+let ManagedTree = React.createClass({
+  propTypes: Tree.propTypes,
+
+  getInitialState() {
+    return { expanded: new WeakMap(),
+             focusedItem: null };
+  },
+
+  setExpanded(item, expanded) {
+    const e = this.state.expanded;
+    e.set(item, expanded);
+    this.setState({ expanded: e });
+
+    if (expanded && this.props.onExpand) {
+      this.props.onExpand(item);
+    } else if (!expanded && this.props.onCollapse) {
+      this.props.onCollapse(item);
+    }
+  },
+
+  focusItem(item) {
+    if (this.state.focused !== item) {
+      this.setState({ focusedItem: item });
+
+      if (this.props.onFocus) {
+        this.props.onFocus(item);
+      }
+    }
+  },
+
+  render() {
+    const { expanded, focusedItem } = this.state;
+
+    const props = Object.assign({}, this.props, {
+      isExpanded: item => expanded.get(item),
```

Comment:
It's essentially a higher-order component. It's quite useful to do this sort of stuff in React, but how we use it definitely should be documented.

The Tree component that we use has no component state (well, right now I think it does track the window height but I want to remove that). It doesn't keep track of the current focused item, or which nodes are expanded. It's all up to the consumer to handle that. It's nice to do it that way for several reasons: it makes tests easier to write, and it allows the consumer to store that info somewhere else like redux if they want to.

But it puts a lot of burden on the user, especially if they are trying to learn how to use the tree. This is a common problem with components: who manages the state? A good technique is to offer two variants of the same component: one that doesn't manage the state, and one that does (here, we are calling it "managed").

So we're just wrapping the Tree widget (so we can't change the prop names, we are passing all of that into the Tree) and handling the necessary state for it to work.

In practice, React has more going on than just "props passed down, events go up". For ease of use it's local component state is really nice. Also, there's really no reason it has to be separated from Redux state. At some point React should allow the consumer to "mount" the internal state of any component somewhere else; you can read about this in this React issue.

Anywhere, we're still figuring out the best patterns here and we'll definitely document them more.